nano/node/bootstrap.cpp

```diff
@@ -3161,3 +3170,67 @@ void nano::frontier_req_server::next ()
 	current = account_pair.first;
 	frontier = account_pair.second;
 }
+
+void nano::pulls_cache::add (nano::pull_info const & pull_a)
+{
+	if (pull_a.processed > 500)
+	{
+		std::lock_guard<std::mutex> guard (pulls_cache_mutex);
+		// Clean old pull
+		if (cache.size () >= max_cache)
+		{
+			cache.erase (cache.begin ());
+		}
+		nano::uint512_union head_512 (pull_a.account, pull_a.head_original);
+		auto existing (cache.get<account_head_tag> ().find (head_512));
+		if (existing == cache.get<account_head_tag> ().end ())
+		{
+			// Insert new pull
+			auto inserted (cache.insert (nano::cached_pulls{ std::chrono::steady_clock::now (), head_512, pull_a.head }));
+			assert (inserted.second);
+		}
+		else
+		{
+			// Update existing pull
+			cache.get<account_head_tag> ().modify (existing, [pull_a](nano::cached_pulls & cache_a) {
+				cache_a.time = std::chrono::steady_clock::now ();
+				cache_a.new_head = pull_a.head;
+			});
+		}
+	}
+}
+
+void nano::pulls_cache::update_pull (nano::pull_info & pull_a)
+{
+	std::lock_guard<std::mutex> guard (pulls_cache_mutex);
+	nano::uint512_union head_512 (pull_a.account, pull_a.head_original);
+	auto existing (cache.get<account_head_tag> ().find (head_512));
+	if (existing != cache.get<account_head_tag> ().end ())
+	{
+		pull_a.head = existing->new_head;
+	}
+}
+
+void nano::pulls_cache::remove (nano::pull_info const & pull_a)
+{
+	std::lock_guard<std::mutex> guard (pulls_cache_mutex);
+	nano::uint512_union head_512 (pull_a.account, pull_a.head_original);
+	cache.get<account_head_tag> ().erase (head_512);
+}
+
+namespace nano
+{
+std::unique_ptr<seq_con_info_component> collect_seq_con_info (pulls_cache & pulls_cache, const std::string & name)
```

Comment:
This should be added to the `collect_seq_con_info (bootstrap_initiator &, const std::string &)` function:

`composite->add_component (collect_seq_con_info (bootstrap_initiator.cache, "cache"));`

otherwise it isn't be used.